File: spC++/src/Implement/Services/BookStatistics.cpp
#include "Headers/Services/BookStatistics.hpp"

void BookStatistics::visitBook(Book* book) {
    for (Visitee* element : book->getChildren()) {
        element->accept(this);
    }
}

void BookStatistics::visitSection(Section* section) {
    for (IElement* element : section->getChildren()) {
        element->accept(this);
    }
}

void BookStatistics::visitToC(TableOfContents* tableOfContents) {
    return ;
}

void BookStatistics::visitParagraph(Paragraph* paragraph) {
    this->NoParagraphs++;
}

void BookStatistics::visitIProxy(ImageProxy* imageProxy) {
    imageProxy->loadImage().accept(this);
}

void BookStatistics::visitImage(Image* image) {
    this->NoImages++;
}

void BookStatistics::visitTable(Table* table) {
    this->NoTables++;
}

void BookStatistics::visitAuthor(Author* author) {
    return ;
}

void BookStatistics::printStatistics() {
    std::cout << "Book statistics: " << std::endl << std::endl;
    std::cout << "Number of images: " << this->NoImages << std::endl;
    std::cout << "Number of tables: " << this->NoTables << std::endl;
    std::cout << "Number of paragraphs: " << this->NoParagraphs << std::endl;
}

File: Java-Labs/lab/src/main/java/sp/lab/Services/BookStatistics.java
package sp.lab.Services;

import sp.lab.Models.Book;
import sp.lab.Models.IVisitor;
import sp.lab.Models.Image;
import sp.lab.Models.ImageProxy;
import sp.lab.Models.Paragraph;
import sp.lab.Models.Section;
import sp.lab.Models.Table;
import sp.lab.Models.TableOfContents;

public class BookStatistics implements IVisitor{
    private int images, tables, paragraphs;

    public BookStatistics() {
        this.images = 0;
        this.tables = 0;
        this.paragraphs = 0;
    }

    @Override
    public void visitBook(Book book) {}

    @Override
    public void visitSection(Section section) {}

    @Override
    public void visitParagraph(Paragraph paragraph) {
        this.paragraphs++;
    }

    @Override
    public void visitImageProxy(ImageProxy imageProxy) {
        this.images++;
    }

    @Override
    public void visitImage(Image image) {
        this.images++;
    }

    @Override
    public void visitTable(Table table) {
        this.tables++;
    }

    @Override
    public void visitTableOfContents(TableOfContents tableOfContents) {}

    public void printStatistics() {
        System.out.println("Book statistics: ");
        System.out.println("\tImages: " + this.images);
        System.out.println("\tTables: " + this.tables);
        System.out.println("\tParagraphs: " + this.paragraphs);
    }
}
